### PR TITLE
fix: validate non-signed commit fields against `LastProposedCommit`

### PR DESCRIPTION
**Pull request by aljo242**
Closes BLO-994
Closes BLO-995
Closes BLO-976
Closes BLO-1001
Closes BLO-951

Add a check for consistency between the commit abci requestProcessProposal info and the extended commit info.

Checks:
- validator address
- validator power
- blockIDFlag
- consensus round

clean up tests and add additional checks

### CODE REVIEW

File: abci/proposals/validate.go [hunk 1]
```diff
@@ -1,17 +1,19 @@
 package proposals
 
 import (
+	"fmt"
+
 	cometabci "github.com/cometbft/cometbft/abci/types"
 	sdk "github.com/cosmos/cosmos-sdk/types"
 
 	"github.com/skip-mev/slinky/abci/ve"
 )
```
Comment (g3ndo9kar9) [inline abci/proposals/validate.go]: Why separate this logic between Prepare / Process? We need to ensure parity between Prepare / Process, so separating validation into separate functions is a foot-gun IMO.

Comment (g3ndo9kar9) [reply abci/proposals/validate.go]: IMO there should be a single criteria we're always using here

File: abci/proposals/validate.go [hunk 2]
```diff
@@ -58,3 +60,116 @@ func (h *ProposalHandler) ValidateExtendedCommitInfo(
...
+// marshalled and contain valid prices. This function contains extra validation to be run in
+// ProcessProposal.
+// CONTACT: this function assumes that the RequestProcessProposal is not nil and has already been
+// checked by the caller.
+func (h *ProposalHandler) ValidateExtendedCommitInfoProcess(
+	ctx sdk.Context,
+	req *cometabci.RequestProcessProposal,
+	extendedCommitInfo cometabci.ExtendedCommitInfo,
+) error {
+	if err := h.validateVoteExtensionsFn(ctx, extendedCommitInfo); err != nil {
```
Comment (g3ndo9kar9) [inline abci/proposals/validate.go]: let's just fix this as the sdk.ValidateVoteExtensions function

File: abci/proposals/validate.go [hunk 3]
```diff
@@ -58,3 +60,116 @@ func (h *ProposalHandler) ValidateExtendedCommitInfo(
...
+			"failed to validate vote extensions; vote extensions may not comprise a supermajority",
+			"height", req.Height,
+			"err", err,
+		)
+
+		return err
+	}
+
+	// checks for consistency between extendedCommitInfo and proposedLastCommit
+	if extendedCommitInfo.Round != req.ProposedLastCommit.Round {
```
Comment (g3ndo9kar9) [inline abci/proposals/validate.go]: This check is performed in the SDK's validate VE function

File: abci/proposals/validate.go [hunk 4]
```diff
@@ -58,3 +60,116 @@ func (h *ProposalHandler) ValidateExtendedCommitInfo(
...
+			"proposed last commit round", req.ProposedLastCommit.Round,
+		)
+
+		return fmt.Errorf("mismatched round in encoded extended commit info and proposed last commit: %d != %d",
+			extendedCommitInfo.Round,
+			req.ProposedLastCommit.Round,
+		)
+	}
+
+	if len(extendedCommitInfo.Votes) != len(req.ProposedLastCommit.Votes) {
```
Comment (g3ndo9kar9) [inline abci/proposals/validate.go]: same as above

File: abci/proposals/validate.go [hunk 5]
```diff
@@ -58,3 +60,116 @@ func (h *ProposalHandler) ValidateExtendedCommitInfo(
...
+			h.logger.Error(
+				"no vote for validator in extended commit vote found in proposed last commit",
+				"height", req.Height,
+				"validator", string(address),
+			)
+
+			return fmt.Errorf("no vote for validator in extended commit vote found in proposed last commit")
+		}
+
+		if reqVote.Validator.Power != vote.Validator.Power {
```
Comment (g3ndo9kar9) [inline abci/proposals/validate.go]: This check is also perfomed in the upstream ValidateVoteExtensions fn handler

File: abci/proposals/validate.go [hunk 6]
```diff
@@ -58,3 +60,116 @@ func (h *ProposalHandler) ValidateExtendedCommitInfo(
...
+				"height", req.Height,
+				"validator", string(address),
+				"extended vote power", vote.Validator.Power,
+				"last proposed vote power", reqVote.Validator.Power,
+			)
+
+			return fmt.Errorf("mismatched validator power between extended commit vote and last proposed commit")
+		}
+
+		if reqVote.BlockIdFlag != vote.BlockIdFlag {
```
Comment (g3ndo9kar9) [inline abci/proposals/validate.go]: I looked into this again, this actually is correct, the proposedBlock.LastCommit has each vote's BlockFlagID

Comment (g3ndo9kar9) [reply abci/proposals/validate.go]: I'd prefer that we just add this check to the `ValidateVoteExtensions` from the SDK (and move the upstream impl into our codebase)

File: abci/ve/utils.go [hunk 7]
```diff
@@ -85,3 +93,182 @@ func NoOpValidateVoteExtensions(
...
+type ValidatorStore interface {
+	GetPubKeyByConsAddr(context.Context, sdk.ConsAddress) (cmtprotocrypto.PublicKey, error)
+}
+
+// ValidateVoteExtensions defines a helper function for verifying vote extension
+// signatures that may be passed or manually injected into a block proposal from
+// a proposer in PrepareProposal. It returns an error if any signature is invalid
+// or if unexpected vote extensions and/or signatures are found or less than 2/3
+// power is received.
+func ValidateVoteExtensions(
```
Comment (g3ndo9kar9) [inline abci/ve/utils.go:113]: can we pull in the test-cases from the SDK as well? They were implemented as a separate test-suite from base-app so shld be low effort (just copy / paste)

File: providers/base/websocket/handlers/mocks/web_socket_conn_handler.go
Comment (g3ndo9kar9) [inline providers/base/websocket/handlers/mocks/web_socket_conn_handler.go:1]: Why is this rebuilt?

File: config/local/oracle.json
Comment (g3ndo9kar9) [inline config/local/oracle.json:1]: Why is this changed?

File: providers/base/websocket/handlers/mocks/web_socket_conn_handler.go
Comment (g3ndo9kar9) [reply providers/base/websocket/handlers/mocks/web_socket_conn_handler.go:1]: Same for all other mocks in websocket

File: abci/ve/README.md [hunk 8]
```diff
@@ -4,14 +4,14 @@
 
 ## Overview
 
-Each validator in the network running the slinky module either runs an oracle in process or out of process. In either case, the oracle is responsible for fetching data offchain that will included the validators vote extension, broadcasted to the network, included in a block, and subsequently committed on-chain.
+Each validator in the network running the slinky module either runs an oracle in process or out of process. In either case, the oracle is responsible for fetching data offchain that will include the validator's vote extension, broadcast to the network, included in a block, and subsequently committed on-chain.
```
Comment (g3ndo9kar9) [inline abci/ve/README.md:8]: kek

File: providers/base/websocket/handlers/mocks/web_socket_conn_handler.go
Comment (aljo242) [reply providers/base/websocket/handlers/mocks/web_socket_conn_handler.go:1]: I have no idea lol exclude them

File: config/local/oracle.json
Comment (aljo242) [reply config/local/oracle.json:1]: I must have run some `mock` generation that generated everything

File: abci/preblock/oracle/preblock.go [hunk 9]
```diff
@@ -112,7 +112,7 @@ func (h *PreBlockHandler) PreBlocker() sdk.PreBlocker {
 		}()
 
 		// If vote extensions are not enabled, then we don't need to do anything.
-		if !ve.VoteExtensionsEnabled(ctx) || req == nil {
```
Comment (g3ndo9kar9) [inline abci/preblock/oracle/preblock.go:115]: Why remove nil request check?

File: abci/ve/utils.go [hunk 10]
```diff
@@ -85,3 +97,186 @@ func NoOpValidateVoteExtensions(
...
+	addressCache := make(map[string]struct{}, len(ec.Votes))
+	// check that consistency between LastCommit and ExtendedCommit
+	for i, vote := range ec.Votes {
+		// cache addresses to check for duplicates
+		if _, ok := addressCache[string(vote.Validator.Address)]; ok {
+			return fmt.Errorf("extended commit vote address %X is duplicated", vote.Validator.Address)
+		}
+		addressCache[string(vote.Validator.Address)] = struct{}{}
+
+		if !bytes.Equal(vote.Validator.Address, lc.Votes().Get(i).Validator().Address()) {
```
Comment (g3ndo9kar9) [inline abci/ve/utils.go]: nit: make a temp var for the vote in last commit instead of indexing each time it's used

**Review by g3ndo9kar9 — APPROVED** [review]
lgtm

File: abci/preblock/oracle/preblock.go (hunk 9 above)
Comment (aljo242) [reply abci/preblock/oracle/preblock.go:115]: we check nil earlier in the func
